### PR TITLE
adapter: support explicit single statement transactions

### LINKED ISSUE #20426: Support single DDL in explicit transactions
## Summary

Allow explicit transactions that contain a single DDL statement, like:

```sql
BEGIN;
CREATE TABLE t1 (a int);
COMMIT;
```

This will unblock running DDL in at least:

  * Deepnote
  * DBeaver

## Details

Materialize today does not support executing DDL transactionally. E.g., you can't create two tables in a single transaction and have Materialize atomically create both tables or neither:

```sql
BEGIN;
CREATE TABLE t1 (a int);
CREATE TABLE t2 (a int);
COMMIT;
```

However, Materialize's mechanism of enforcement is quite restrictive. Materialize rejects any explicit transaction that contains even a single DDL statement:

```sql
=> BEGIN;
=> CREATE TABLE test (a int);
ERROR:  CREATE TABLE test (a int4) cannot be run inside a transaction block
```

I don't _think_ there is any technical reason that we couldn't allow exactly one DDL statement in an explicit transaction block.

DBeaver always executes DDL statements in an explicit transaction block, and so currently always fails like so when attempting to create or alter the definition of any SQL objects:

Deepnote similarly has the same issue:

image

I imagine there are other SQL tools that suffer from the same problem.

## Implementation notes

  * Keep #9999 in mind. We want to be sure we don't accidentally allow DDL in transactions that do anything besides DDL.

cc @mjibson

### PR DESCRIPTION
**Pull request by mjibson**
See included design doc. Add a new transaction operation for delayed statement execution.

Fixes #20426

### Motivation

  * This PR adds a known-desirable feature.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - Support most single statements in explicit transactions: `BEGIN; CREATE <x>; COMMIT;`.

### CODE REVIEW

File: doc/developer/design/20230720_single_statement_explitic_transaction.md [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+
+#### Changes from this design
+
+This design adds a new operation to the inner transaction: `SingleStatement`.
+This operation records a single statement that is later retrieved and executed during `COMMIT`.
+
+This design also adds what is essentially a new handler: the coordinator itself.
+When a transaction is being committed and its inner operation is `SingleStatement`, the inner statement is queued in the coordinator for execution by a new method that is serving as a handler: `sequence_execute_single_statement_transaction`.
+This method:
+1. Ensures the session's transaction is in `Default`, which should be a side effect of running it through `sequence_end_transaction` from the `COMMIT`.
```
Comment (jkosh44) [inline doc/developer/design/20230720_single_statement_explitic_transaction.md:85]: I was finding this a bit confusing as I read through the code, it seems like the status really should be in the `InTransaction` state because it's in an explicit transaction that hasn't committed yet. The fact that this transaction goes through two cycles of default->in-transaction->commit is a bit hard to follow and reason about.

File: doc/developer/design/20230720_single_statement_explitic_transaction.md [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+2. Puts the session's transaction into `Started` (single-statement, implicit transaction).
```
Comment (jkosh44) [inline doc/developer/design/20230720_single_statement_explitic_transaction.md:86]: Similarly it was a bit confusing to me that we set the state to an implicit transaction when the transaction is an explicit transaction. It sounds like we're just trying to take advantage of some nice properties that implicit transactions have but it's not really an implicit transaction? 

If so I think it would be clearer to add a new state that has those properties but is specific to these single statement transactions. Otherwise whenever someone makes changes to the implicit transaction state they'll have to remember that they're also making changes to the single statement transaction state, and it seems easy to forget about this.

File: doc/developer/design/20230720_single_statement_explitic_transaction.md [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+
+## Detailed description
+
+Many statements cannot be executed without observable side effects to other sessions, for example all DDLs.
+However, these and many other statements do have a known `ExecuteResponse` that can be determined without execution (as opposed to statements that return a row count).
+We will add a new transaction operation mode called `SingleStatement`.
+Currently those modes are `Peeks`, `Writes`, and `Subscribe`.
+`SingleStatement` will record the statement, then generate and return to the user an `ExecuteResponse` that assummes success.
+This mode can only be entered in explicit transactions (must start with `BEGIN`).
+If the next statement is `COMMIT`, the coordinator sets the session's transaction status to `InTransactionImplicit`, which is a status that can execute any statement, then executes the statement.
```
[the remark this answers is not in the document]
Comment (mjibson) [reply doc/developer/design/20230720_single_statement_explitic_transaction.md]: This should have been `Started` instead of `InTransactionImplicit`, although that won't resolve your question. Yes, that's what happens, but with a few other steps that I've added. The explicit COMMIT causes the transaction to clear and process its inner operation. If that operation is `SingleStatement`, then we start a new txn, run the stmt, run commit. I've added text for all of this to clarify.

File: doc/developer/design/20230720_single_statement_explitic_transaction.md (hunk 1 above)
Comment (mjibson) [reply doc/developer/design/20230720_single_statement_explitic_transaction.md]: I agree it's weird because of the extra cycle. This feature is essentially a new transaction operation: `SingleStatement`. The other operations (`Peeks`, "Writes`) already do things *after* the transaction has cleared but during "operation processing" time. Peeks can ship off to the strict serializable wait, Writes can ship off to PendingWrites. `SingleStatement`'s operation processing is to enqueue a new coord message that executes the recorded statement. I've added some text to explain more.

File: doc/developer/design/20230720_single_statement_explitic_transaction.md (hunk 2 above)
Comment (mjibson) [reply doc/developer/design/20230720_single_statement_explitic_transaction.md:88]: I'm not sure doing this helps us much. We would still need to teach sequence_end_transaction to detect this status and it'd then still have to enqueue a message to do the work, and the only way to do the work is to call handle_execute, which is the thing doing all the transaction status checking, forcing the manual starting of an implicit transaction. We could refactor handle_execute to split up those parts and skip the checking, but I don't think it's safer.

Also, note way above I incorrectly used `InTransactionImplicit` (multi statement implicit) where it should have been `Started` (single statement implicit). `Started` is pretty clear about what's going on, and seems like a safe status to use.

I think my summary understanding of your idea (a new status) is: it's not clear to me how we would use that status to both delay execution and then later actually execute.

File: src/adapter/src/coord/command_handler.rs [hunk 4]
```diff
@@ -594,21 +620,22 @@ impl Coordinator {
...
                 task::spawn(|| format!("purify:{conn_id}"), async move {
                     let catalog = catalog.for_session(ctx.session());
+
+                    // Checks if the session is authorized to purify a statement. Usually
+                    // authorization is checked after planning, however purification happens before
+                    // planning, which may require the use of some connections and secrets.
+                    if let Err(e) = rbac::check_item_usage(&catalog, ctx.session(), &resolved_ids) {
+                        return ctx.retire(Err(e));
+                    }
+
```
[the remark this answers is not in the document]
Comment (mjibson) [reply src/adapter/src/coord/command_handler.rs:638]: Reusing the `catalog` object from way up above caused a rust borrow issue and this needed the new one.

File: src/adapter/src/error.rs [hunk 5]
```diff
@@ -399,6 +401,7 @@ impl AdapterError {
             AdapterError::PreparedStatementExists(_) => SqlState::DUPLICATE_PSTATEMENT,
             AdapterError::ReadOnlyTransaction => SqlState::READ_ONLY_SQL_TRANSACTION,
             AdapterError::ReadWriteUnavailable => SqlState::INVALID_TRANSACTION_STATE,
+            AdapterError::SingleStatementTransaction => SqlState::READ_ONLY_SQL_TRANSACTION,
```
[the remark this answers is not in the document]
Comment (mjibson) [reply src/adapter/src/error.rs]: Agree, copy paste error.

File: src/adapter/src/coord/command_handler.rs [hunk 6]
```diff
@@ -531,47 +532,72 @@ impl Coordinator {
...
+                        // be run in a special single-statement explicit mode. In this mode (`BEGIN;
+                        // <stmt>; COMMIT`), we generate the expected tag from a successful <stmt>, but
+                        // delay execution until `COMMIT`.
+                        let mut resp_kind = Plan::generated_from((&stmt).into())
+                            .into_iter()
+                            .map(ExecuteResponse::generated_from)
+                            .flatten()
+                            .map(|r| r.try_into())
+                            .collect::<Vec<Result<ExecuteResponse, _>>>();
+                        // If we're in an implicit transaction and we could generate exactly one
```
[the remark this answers is not in the document]
Comment (mjibson) [reply src/adapter/src/coord/command_handler.rs]: yes

File: src/adapter/src/coord/sequencer.rs [hunk 7]
```diff
@@ -518,6 +520,65 @@ impl Coordinator {
...
+    pub(crate) async fn sequence_execute_single_statement_transaction(
+        &mut self,
+        ctx: ExecuteContext,
+        stmt: Statement<Raw>,
+        params: Params,
+    ) {
+        // Put the session into single statement implicit so anything can execute.
+        let (tx, internal_cmd_tx, mut session, extra) = ctx.into_parts();
+        assert!(matches!(session.transaction(), TransactionStatus::Default));
+        session.start_transaction_implicit(self.now_datetime(), 1);
```
Comment (jkosh44) [inline src/adapter/src/coord/sequencer.rs:533]: I think this is the functional call that confused me. Can we add a different function that only ever enters the `Started` state? Something like `start_transaction_single_stmt`.

File: doc/developer/design/20230720_single_statement_explitic_transaction.md (hunk 2 above)
Comment (mjibson) [reply doc/developer/design/20230720_single_statement_explitic_transaction.md:88]: Done.

File: src/adapter/src/coord/sequencer.rs (hunk 7 above)
Comment (mjibson) [reply src/adapter/src/coord/sequencer.rs]: Done.

**Comment by mjibson** [comment]
There were a few tests that asserted some statements could not be executed in transactions. Those test lines have been removed.

**Review by jkosh44 — APPROVED** [review]

File: doc/developer/design/20230720_single_statement_explitic_transaction.md [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+
+Many statements cannot be executed without observable side effects to other sessions, for example all DDLs.
+However, these and many other statements do have a known `ExecuteResponse` that can be determined without execution (as opposed to statements that return a row count).
+We will add a new transaction operation mode called `SingleStatement`.
+Currently those modes are `Peeks`, `Writes`, and `Subscribe`.
+`SingleStatement` will record the statement, then generate and return to the user an `ExecuteResponse` that assummes success.
+The `SingleStatement` operation can only be entered in explicit transactions (must start with `BEGIN`).
+On `COMMIT`, the coordinator clears the current transaction, fetching and processing its inner operation (this already occurs before this document).
+If the inner operation is `SingleStatement`, the transaction status is set to `Started` (a status that can execute any statement), the recorded statement is executed, and an implicit `COMMIT` is executed.
+The implicit `COMMIT` puts the transaction into the same final state (`Default`) as the explicit `COMMIT`.
```
Comment (benesch) [inline doc/developer/design/20230720_single_statement_explitic_transaction.md:42]: Could you briefly explain here what happens if a user runs `BEGIN;` then `CREATE ...`; then `SELECT ...`? I took a brief skim of the code and it looks like we properly reject that `SELECT` statement, but would be great to have a sentence or two here explaining the check.

**Review by benesch — APPROVED** [review]
I only read through the design doc. The design doc was really nice! Thank you for writing it. I'm short on time, so I didn't have bandwidth to fully explore the edge cases, but at a high level the protocol you described exactly matched my expectations for how this would work.

Really excited to see this ship. Gonna be a huge win for DBeaver and Deepnote, and surely a dozen other tools we're not explicitly tracking.

File: doc/developer/design/20230720_single_statement_explitic_transaction.md (hunk 8 above)
Comment (mjibson) [reply doc/developer/design/20230720_single_statement_explitic_transaction.md:42]: Will add in follow-up, want this to get in the release.
